exercises/practice/sgf-parsing/.meta/src/SgfParsing.example.elm

```diff
@@ -0,0 +1,144 @@
+module SgfParsing exposing (Tree(..), parse)
+
+import Dict exposing (Dict)
+import Parser exposing ((|.), (|=), DeadEnd, Parser, Problem(..), Step(..), Trailing(..))
+
+
+type Tree
+    = Node { properties : Dict String (List String), children : List Tree }
+
+
+parse : String -> Result String Tree
+parse =
+    Parser.run (sgfWithParentheses |. Parser.end)
+        >> Result.mapError
+            (List.reverse >> List.head >> Maybe.map deadEndsToString >> Maybe.withDefault "")
+
+
+deadEndsToString : DeadEnd -> String
+deadEndsToString { problem } =
+    case problem of
+        ExpectingSymbol "(" ->
+            "tree missing"
+
+        ExpectingSymbol ";" ->
+            "tree with no nodes"
+
+        Expecting "[" ->
+            "properties without delimiter"
+
+        Problem string ->
+            string
+
+        _ ->
+            "Unexpected error"
+
+
+sgf : Parser Tree
+sgf =
+    Parser.oneOf [ sgfWithParentheses, sgfWithoutParentheses ]
+
+
+sgfWithParentheses : Parser Tree
+sgfWithParentheses =
+    Parser.succeed identity
+        |. Parser.symbol "("
+        |= Parser.lazy (\_ -> sgfWithoutParentheses)
+        |. Parser.symbol ")"
+
+
+sgfWithoutParentheses : Parser Tree
+sgfWithoutParentheses =
+    Parser.succeed (\prop trees -> Node { properties = prop, children = trees })
+        |. Parser.symbol ";"
+        |= Parser.loop [] properties
```

Comment:
I think loop is less common than sequence, and the syntax of how it works looks a little strange to me. It feels a bit like an advanced topic.

It probably is better though with the tail call elimination, and other examples will be submitted over time anyway, so probably good to keep it as it is.

Reply:
I actually started with a sequence, and I think it worked, but all the properties (`start`, `stop`...) except for `item` were empty or irrelevant so it felt strange, especially using `spaces` felt like a waste of resources.
I agree with you that I would not use that for teaching about parsing, but this imoplementation is just meant as a proof that the exercise can be solved.